### PR TITLE
[PatternMatch] Do not accept undef elements in m_AllOnes() and friends

### PR DESCRIPTION
**Pull request by nikic**
Change all the cstval_pred_ty based PatternMatch helpers (things like m_AllOnes and m_Zero) to only allow poison elements inside vector splats, not undef elements.

Historically, we used to represent non-demanded elements in vectors using undef. Nowadays, we use poison instead. As such, I believe that support for undef in vector splats is no longer useful.

At the same time, while poison splat elements are pretty much always safe to ignore, this is not generally the case for undef elements. We have existing miscompiles in our tests due to this (see the masked-merge-*.ll tests changed here) and it's easy to miss such cases in the future, now that we write tests using poison instead of undef elements.

I think overall, keeping support for undef elements no longer makes sense, and we should drop it. Once this is done consistently, I think we may also consider allowing poison in m_APInt by default, as doing that change is much less risky than doing the same with undef.

This PR is a stub for discussion -- this change has more than a hundred test failures in InstCombine alone, and I don't want to update all of them before we reach a consensus. The abs-1.ll test is a representative example of how I would update most tests: By replacing undef with poison. I don't think there is value in retaining the undef test coverage anymore at this point.

### CODE REVIEW

**Review by nunoplopes — APPROVED** [review]
I think this is great, thank you!
There have been cases of wrong optimizations because of the match of undef elements, so this should fix a few of such cases.

**Review by dtcxzyw — APPROVED** [review]
LGTM. BTW, can we drop the call to `Constant::replaceUndefsWith` in InstCombine after this patch lands?
See also https://github.com/llvm/llvm-project/pull/85772/files#r1530496643.

**Comment by nikic** [comment]
@dtcxzyw It will take more changes to get there, but yes, we should definitely get rid of such replaceUndefsWith calls.

**Comment by nikic** [comment]
Okay, this should be ready for another round of review now. I've updated all the affected tests now (mostly by replacing undef -> poison, in some cases by duplicating tests where I thought it had value).

I've also made some minor code cleanup while doing the test updates. The main one is that I dropped the m_NotForbidUndef matcher, because all uses of it are fine with poison elements.

File: llvm/test/Transforms/InstSimplify/shift.ll [hunk 1]
```diff
@@ -306,11 +306,22 @@ define <2 x i7> @all_ones_left_right_splat(<2 x i7> %x) {
...
+  %right = ashr <3 x i7> %left, %x
+  ret <3 x i7> %right
+}
+
+define <3 x i7> @all_ones_left_right_splat_poison__elt(<3 x i7> %x) {
+; CHECK-LABEL: @all_ones_left_right_splat_poison__elt(
 ; CHECK-NEXT:    ret <3 x i7> <i7 -1, i7 -1, i7 -1>
 ;
-  %left = shl <3 x i7> <i7 poison, i7 -1, i7 undef>, %x
+  %left = shl <3 x i7> <i7 poison, i7 -1, i7 poison>, %x
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstSimplify/shift.ll:324]: Was the original transform bad?

Comment (nikic) [reply llvm/test/Transforms/InstSimplify/shift.ll:324]: No, the transform was fine as it did not propagate the undef. This test corresponds to the `// We could return the original -1 constant to preserve poison elements.` comment change in InstructionSimplify.cpp: Now that we only handle poison elements, we *could* propagate the poison elements. I didn't make the actual change for that in this PR.

File: llvm/test/Transforms/InstCombine/add.ll [hunk 2]
```diff
@@ -2983,7 +2983,7 @@ define i8 @signum_i8_i8_use3(i8 %x) {
   ret i8 %r
 }
 
-; poison/undef is ok to propagate in shift amount
+; poison/poison is ok to propagate in shift amount
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/add.ll:2986]:
```suggestion
; poison is ok to propagate in shift amount
```

File: llvm/test/Transforms/InstCombine/and-or-icmps.ll [hunk 3]
```diff
@@ -965,14 +965,29 @@ define <2 x i1> @substitute_constant_or_ne_slt_swap_vec(<2 x i8> %x, <2 x i8> %y
   ret <2 x i1> %r
 }
 
+; The poison case would be valid to fold.
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/and-or-icmps.ll:968]:
```suggestion
; TODO: The poison case would be valid to fold.
```

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by nunoplopes** [comment]
Alive2 is complaining about one case (only!):
```llvm
; Transforms/InstCombine/integer-round-up-pow2-alignment.ll

define <2 x i4> @t18_replacement_0b0001(<2 x i4> %x) {
  %x.lowbits = and <2 x i4> %x, { 3, 3 }
  %x.lowbits.are.zero = icmp eq <2 x i4> %x.lowbits, { 0, 0 }
  %x.biased = add <2 x i4> %x, { 3, 3 }
  %x.biased.highbits = and <2 x i4> %x.biased, { 12, poison }
  call void @use.v2i4(<2 x i4> %x.biased.highbits)
  %x.roundedup = select <2 x i1> %x.lowbits.are.zero, <2 x i4> %x, <2 x i4> %x.biased.highbits
  ret <2 x i4> %x.roundedup
}
=>
define <2 x i4> @t18_replacement_0b0001(<2 x i4> %x) {
  %x.biased = add <2 x i4> %x, { 3, 3 }
  %x.biased.highbits = and <2 x i4> %x.biased, { 12, poison }
  call void @use.v2i4(<2 x i4> %x.biased.highbits)
  ret <2 x i4> %x.biased.highbits
}
Transformation doesn't verify! (unsound)
ERROR: Target is more poisonous than source

Example:
<2 x i4> %x = < #x0 (0), #x0 (0) >

Source:
<2 x i4> %x.lowbits = < #x0 (0), #x0 (0) >
<2 x i1> %x.lowbits.are.zero = < #x1 (1), #x1 (1) >
<2 x i4> %x.biased = < #x3 (3), #x3 (3) >
<2 x i4> %x.biased.highbits = < #x0 (0), poison >
Function @use.v2i4 returned
<2 x i4> %x.roundedup = < #x0 (0), #x0 (0) >

Target:
<2 x i4> %x.biased = < #x3 (3), #x3 (3) >
<2 x i4> %x.biased.highbits = < #x0 (0), poison >
Function @use.v2i4 returned
Source value: < #x0 (0), #x0 (0) >
Target value: < #x0 (0), poison >
```

Several tests got fixed btw! 🙏

**Comment by goldsteinn** [comment]
Think for this to be consistent also need #88217

**Comment by nikic** [comment]
@nunoplopes Thanks, this should be fixed by https://github.com/llvm/llvm-project/commit/525d00e5edc4b83105c6ad518926b174f76c3a1c.

**Pull request merged**